File: app/src/main/java/com/gnoemes/shikimoriapp/presentation/view/alternative/translations/AlternativeTranslationsFragment.java
package com.gnoemes.shikimoriapp.presentation.view.alternative.translations;

import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.afollestad.materialdialogs.MaterialDialog;
import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;
import com.gnoemes.shikimoriapp.R;
import com.gnoemes.shikimoriapp.entity.anime.series.domain.AlternativeTranslationType;
import com.gnoemes.shikimoriapp.entity.anime.series.presentation.AlternativeTranslationNavigationData;
import com.gnoemes.shikimoriapp.entity.anime.series.presentation.AlternativeTranslationViewModel;
import com.gnoemes.shikimoriapp.entity.app.presentation.AppExtras;
import com.gnoemes.shikimoriapp.presentation.presenter.alternative.AlternativeTranslationsPresenter;
import com.gnoemes.shikimoriapp.presentation.view.alternative.translations.adapter.TranslationAdapter;
import com.gnoemes.shikimoriapp.presentation.view.common.fragment.BaseFragment;
import com.gnoemes.shikimoriapp.presentation.view.common.fragment.RouterProvider;
import com.gnoemes.shikimoriapp.presentation.view.common.widget.EmptyContentViewWithButton;
import com.gnoemes.shikimoriapp.presentation.view.common.widget.NetworkErrorView;
import com.gnoemes.shikimoriapp.utils.view.DrawableHelper;
import com.gnoemes.shikimoriapp.utils.view.VerticalSpaceItemDecoration;

import java.util.Arrays;
import java.util.List;

import butterknife.BindView;

public class AlternativeTranslationsFragment extends BaseFragment<AlternativeTranslationsPresenter, AlternativeTranslationsView>
        implements AlternativeTranslationsView {

    @BindView(R.id.view_empty)
    EmptyContentViewWithButton emptyContentView;

    @BindView(R.id.list)
    RecyclerView recyclerView;

    @BindView(R.id.refresh_layout)
    SwipeRefreshLayout refreshLayout;

    @BindView(R.id.view_network_error)
    NetworkErrorView networkErrorView;

    @InjectPresenter
    AlternativeTranslationsPresenter presenter;
    private TranslationAdapter adapter;

    public static AlternativeTranslationsFragment newInstance(AlternativeTranslationNavigationData data) {
        AlternativeTranslationsFragment fragment = new AlternativeTranslationsFragment();
        Bundle args = new Bundle();
        args.putLong(AppExtras.ARGUMENT_EPISODE_ID, data.getEpisodeId());
        args.putSerializable(AppExtras.ARGUMENT_TRANSLATION_TYPE, data.getType());
        args.putLong(AppExtras.ARGUMENT_ANIME_ID, data.getAnimeId());
        args.putLong(AppExtras.ARGUMENT_ANIME_RATE_ID, data.getRateId());
        fragment.setArguments(args);
        return fragment;
    }

    @ProvidePresenter
    AlternativeTranslationsPresenter providePresenter() {
        presenter = presenterProvider.get();
        if (getArguments() != null) {
            if (getParentFragment() != null) {
                presenter.setLocalRouter(((RouterProvider) getParentFragment()).getLocalRouter());
            }
            presenter.setEpisodeId(getArguments().getLong(AppExtras.ARGUMENT_EPISODE_ID));
            presenter.setCurrentTranslationType((AlternativeTranslationType) getArguments().getSerializable(AppExtras.ARGUMENT_TRANSLATION_TYPE));
            presenter.setAnimeId(getArguments().getLong(AppExtras.ARGUMENT_ANIME_ID));
            presenter.setRateId(getArguments().getLong(AppExtras.ARGUMENT_ANIME_RATE_ID));
        }

        return presenter;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initViews();
    }

    private void initViews() {

        emptyContentView.setText(R.string.translations_empty);
        emptyContentView.setCallback(v -> getPresenter().onFindAll());
        emptyContentView.setVisibility(View.GONE);
        networkErrorView.setVisibility(View.GONE);

        refreshLayout.setOnRefreshListener(() -> getPresenter().loadTranslations());

        adapter = new TranslationAdapter(getPresenter()::onTranslationClicked);

        int margin = (int) getResources().getDimension(R.dimen.margin_small);
        recyclerView.addItemDecoration(new VerticalSpaceItemDecoration(margin));
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(adapter);

        Drawable navigationIcon = DrawableHelper
                .withContext(getContext())
                .withDrawable(R.drawable.ic_arrow_back)
                .withAttributeColor(R.attr.colorText)
                .tint()
                .get();

        toolbar.setNavigationIcon(navigationIcon);
        toolbar.setNavigationOnClickListener(v -> getPresenter().onBackPressed());
        toolbar.inflateMenu(R.menu.menu_translations);

        Drawable icon = DrawableHelper
                .withContext(getContext())
                .withDrawable(R.drawable.ic_settings)
                .withAttributeColor(R.attr.colorText)
                .tint()
                .get();

        toolbar.getMenu()
                .getItem(0)
                .setIcon(icon);

        toolbar.setOnMenuItemClickListener(item -> {
            switch (item.getItemId()) {
                case R.id.item_settings:
                    getPresenter().onSettingsClicked();
                    break;
            }
            return false;
        });
    }

    ///////////////////////////////////////////////////////////////////////////
    // GETTERS
    ///////////////////////////////////////////////////////////////////////////

    @Override
    protected AlternativeTranslationsPresenter getPresenter() {
        return presenter;
    }

    @Override
    protected int getFragmentLayout() {
        return R.layout.fragment_translations;
    }

    ///////////////////////////////////////////////////////////////////////////
    // MVP
    ///////////////////////////////////////////////////////////////////////////

    @Override
    public void showTranslations(List<AlternativeTranslationViewModel> models) {
        recyclerView.setVisibility(View.VISIBLE);
        adapter.bindItems(models);
    }

    @Override
    public void hideErrorView() {
        networkErrorView.setVisibility(View.GONE);
    }

    @Override
    public void showErrorView() {
        networkErrorView.setVisibility(View.VISIBLE);
    }

    @Override
    public void showEmptyView() {
        emptyContentView.setVisibility(View.VISIBLE);
        recyclerView.setVisibility(View.GONE);
    }

    @Override
    public void hideEmptyView() {
        emptyContentView.setVisibility(View.GONE);
    }

    @Override
    public void onShowLoading() {
        refreshLayout.setRefreshing(true);
    }

    @Override
    public void onHideLoading() {
        refreshLayout.setRefreshing(false);
    }

    @Override
    public void showSettingsDialog() {
        List<AlternativeTranslationType> types = Arrays.asList(AlternativeTranslationType.values());
        new MaterialDialog.Builder(getContext())
                .items(R.array.translations_type_alternative)
                .itemsCallback((dialog, itemView, position, text) -> {
                    dialog.dismiss();
                    getPresenter().onTypeClicked(types.get(position));
                })
                .autoDismiss(true)
                .titleColorAttr(R.attr.colorText)
                .contentColorAttr(R.attr.colorText)
                .alwaysCallSingleChoiceCallback()
                .backgroundColorAttr(R.attr.colorBackgroundWindow)
                .autoDismiss(false)
                .negativeColorAttr(R.attr.colorAction)
                .negativeText(R.string.common_cancel)
                .onNegative((dialog, which) -> dialog.dismiss())
                .canceledOnTouchOutside(true)
                .build()
                .show();
    }
}

File: app/src/main/java/com/gnoemes/shikimoriapp/presentation/view/alternative/translations/AlternativeTranslationsFragment.kt
package com.gnoemes.shikimoriapp.presentation.view.alternative.translations

import android.Manifest
import android.annotation.SuppressLint
import android.os.Bundle
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import com.afollestad.materialdialogs.MaterialDialog
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.ProvidePresenter
import com.gnoemes.shikimoriapp.R
import com.gnoemes.shikimoriapp.entity.anime.series.domain.AlternativeTranslationType
import com.gnoemes.shikimoriapp.entity.anime.series.presentation.AlternativeTranslationNavigationData
import com.gnoemes.shikimoriapp.entity.anime.series.presentation.AlternativeTranslationViewModel
import com.gnoemes.shikimoriapp.entity.app.presentation.AppExtras
import com.gnoemes.shikimoriapp.presentation.presenter.alternative.AlternativeTranslationsPresenter
import com.gnoemes.shikimoriapp.presentation.view.alternative.translations.adapter.TranslationAdapter
import com.gnoemes.shikimoriapp.presentation.view.alternative.translations.adapter.TranslationItemCallback
import com.gnoemes.shikimoriapp.presentation.view.common.fragment.BaseFragment
import com.gnoemes.shikimoriapp.presentation.view.common.fragment.RouterProvider
import com.gnoemes.shikimoriapp.utils.addBackButton
import com.gnoemes.shikimoriapp.utils.ifNotNull
import com.gnoemes.shikimoriapp.utils.themeDrawable
import com.gnoemes.shikimoriapp.utils.withArgs
import kotlinx.android.synthetic.main.fragment_translations.*
import kotlinx.android.synthetic.main.layout_default_list.*
import permissions.dispatcher.NeedsPermission
import permissions.dispatcher.OnNeverAskAgain
import permissions.dispatcher.OnPermissionDenied

class AlternativeTranslationsFragment : BaseFragment<AlternativeTranslationsPresenter, AlternativeTranslationsView>(),
        AlternativeTranslationsView {

    @InjectPresenter
    lateinit var alternativePresenter: AlternativeTranslationsPresenter

    @ProvidePresenter
    fun providePresenter(): AlternativeTranslationsPresenter {
        alternativePresenter = presenterProvider.get()

        arguments.ifNotNull {

            parentFragment.ifNotNull { parent ->
                alternativePresenter.setLocalRouter((parent as RouterProvider).localRouter)
            }

            alternativePresenter.data = it.getSerializable(AppExtras.ARGUMENT_TRANSLATION_DATA) as AlternativeTranslationNavigationData

        }

        return alternativePresenter
    }

    val translationAdapter: TranslationAdapter by lazy {
        TranslationAdapter(object : TranslationItemCallback {
            override fun onTranslationClicked(translation: AlternativeTranslationViewModel) {
                getPresenter().onTranslationClicked(translation)
            }

            override fun onDownloadTranslation(translation: AlternativeTranslationViewModel) {
                getPresenter().onDownloadTranslation(translation)
            }
        })
    }

    companion object {
        @JvmStatic
        fun newInstance(data: AlternativeTranslationNavigationData) = AlternativeTranslationsFragment()
                .withArgs { putSerializable(AppExtras.ARGUMENT_TRANSLATION_DATA, data) }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViews()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    @SuppressLint("ResourceType")
    private fun initViews() {

        with(viewEmpty) {
            setText(R.string.translations_empty)
            setCallback { getPresenter().onFindAll() }
            visibility = View.GONE
        }
        networkErrorView.visibility = View.GONE
        refreshLayout.setOnRefreshListener { getPresenter().loadTranslations() }

        with(list) {
            itemAnimator = DefaultItemAnimator()
            layoutManager = LinearLayoutManager(context)
            adapter = translationAdapter
        }
        toolbar.addBackButton()
        toolbar.setNavigationOnClickListener { getPresenter().onBackPressed() }
        toolbar.inflateMenu(R.menu.menu_translations)

        toolbar.menu
                .getItem(0).icon = context?.themeDrawable(R.drawable.ic_settings, R.attr.colorText)

        toolbar.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.item_settings -> getPresenter().onSettingsClicked()
            }
            false
        }
    }

    @OnPermissionDenied(Manifest.permission.WRITE_EXTERNAL_STORAGE)
    fun onPermissionDenied() {
        getPresenter().onPermissionDenied()
    }

    @OnNeverAskAgain(Manifest.permission.WRITE_EXTERNAL_STORAGE)
    fun onNeverAskAgain() {
        getPresenter().onNeverAskAgain()
    }

    @NeedsPermission(android.Manifest.permission.WRITE_EXTERNAL_STORAGE)
    fun onPermissionGranted() {
        getPresenter().onPermissionGranted()
    }


    ///////////////////////////////////////////////////////////////////////////
    // GETTERS
    ///////////////////////////////////////////////////////////////////////////

    override fun getPresenter(): AlternativeTranslationsPresenter = alternativePresenter


    override fun getFragmentLayout(): Int = R.layout.fragment_translations

    ///////////////////////////////////////////////////////////////////////////
    // MVP
    ///////////////////////////////////////////////////////////////////////////

    override fun showTranslations(models: MutableList<AlternativeTranslationViewModel>?) {
        list.visibility = View.VISIBLE
        translationAdapter.bindItems(models)
    }

    override fun hideErrorView() {
        networkErrorView.visibility = View.GONE
    }

    override fun showErrorView() {
        networkErrorView.visibility = View.VISIBLE
    }

    override fun showEmptyView() {
        viewEmpty.visibility = View.VISIBLE
        list.visibility = View.GONE
    }

    override fun hideEmptyView() {
        viewEmpty.visibility = View.GONE
    }

    override fun showSettingsDialog() {
        context.ifNotNull {
            MaterialDialog(it).show {
                listItemsSingleChoice(R.array.translations_type_alternative) { _, index, _ -> getPresenter().onTypeClicked(AlternativeTranslationType.values()[index]) }
                negativeButton(R.string.common_cancel)
            }
        }
    }

    override fun checkPermissions() {
//        AlternativeTranslationsFragmentPermissionsDispatcher.onPermissionGrantedWithPermissionCheck(this)
    }
}